sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java

```diff
@@ -1815,6 +2051,69 @@ public void processElement(
 
   @Test
   @Category({RunnableOnService.class, UsesStatefulParDo.class})
+  public void testCombiningStateCoderInference() {
```

Comment:
Nice test coverage for this.

Since coders are often very hard to debug, I think we could use test coverage also for failure cases to make sure the error message is actionable.